### ISSUE TITLE
Index out of bounds when printing large body to HTTP response on Ubuntu in release mode

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#12982]
## Bug Report

Creating on behalf of @jwoertink.

```cr
require "socket"
require "http/server/response"

server = TCPServer.new "0.0.0.0", 8081
spawn do
  if io = server.accept?
    io.sync = false

    r = HTTP::Server::Response.new(io)

    r.print "a" * 45_043
  end
end

client = TCPSocket.new("0.0.0.0", 8081)
client << "message\n"
```

```sh
$ sudo docker run --rm -it -v $PWD:/app -w /app crystallang/crystal:1.7.0 crystal run --release test.cr 
Unhandled exception in spawn: Index out of bounds (IndexError)
  from /usr/share/crystal/src/crystal/spin_lock.cr:25:5 in 'unbuffered_write'
  from /usr/share/crystal/src/http/server/response.cr:263:11 in 'unbuffered_write'
  from test.cr:11:5 in '->'
  from /usr/share/crystal/src/fiber.cr:146:11 in 'run'
  from ??
```

## Reproduction Notes

This one is quite the bear to reproduce. This is the best I got so far, but wanted to get the issue created.

* Only reproduces within an Ubuntu docker image, such as `crystallang/crystal:latest`
  * I also tried `ubuntu:22.04` and manually installed Crystal and same thing
  * I'd also assume it happens on an actual Ubuntu desktop install
  * Doesn't reproduce on mac, or my Arch based desktop
* Only happens in `--release` mode
* Starts happening in Crystal `1.7.0`
  * Versions earlier than that do not reproduce
  * Both `1.6.2` and `1.7.0` are on LLVM `13.0.1`
  * I'll try a bisect to see if i can figure out what commit it started at
* Printing `a` fewer than `45043` times does not produce the error
* Passing a non `IO::Buffered` `IO` type to `HTTP::Server::Response` does not produce the error
  * E.g. `IO::Memory` or `STDOUT`
  * Creating an empty custom `IO` type that includes `IO::Buffered` doesnt change anything either
  * The `IO::Buffered` type must have `.sync = false` and `.flush_on_newline = false` as well to reproduce

## Current Thoughts

Given it only happens with non-synced, when not flushing on new lines, i'm thinking it just has to be something related to https://github.com/crystal-lang/crystal/blob/master/src/io/buffered.cr#L150-L160. Where the size of the response triggering something in one of the one of those like `slice[0, index/count]` calls. E.g. hitting up like https://github.com/crystal-lang/crystal/blob/master/src/slice.cr#L213 which results in the `IndexError`

**Comment by Blacksmoke16** [comment]
Bisect says 6edc9e3920ab6e676bd1a27da084c8e1013cdfc1 is the first bad commit :thinking:.

```text
commit 6edc9e3920ab6e676bd1a27da084c8e1013cdfc1
Author: Dmitri Goutnik <dg@syrec.org>
Date:   Sat Nov 12 06:30:55 2022 -0500

    `Exception::CallStack`: avoid allocations in `LibC.dl_iterate_phdr` (#12625)
    
    Calling `self.read_dwarf_sections` directly from the C callback may lead to reallocations and deadlocks due to the internal lock held by `dl_iterate_phdr` (#10084). Work around this by storing object base address and passing it to `self.read_dwarf_sections` later when `dl_iterate_phdr` returns.

 src/exception/call_stack/elf.cr | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)
```

I mean it's not wrong because it works if you revert that commit on master...

**Comment by straight-shoota** [comment]
https://github.com/crystal-lang/crystal/commit/6edc9e3920ab6e676bd1a27da084c8e1013cdfc1 looks very odd. But I can confirm all these findings.

I think this commit has some kind of side effect on the generated code *outside* of the code that's actually affected by this change. Running with `CRYSTAL_LOAD_DEBUG_INFO=0` should completely skip execution of `load_debug_info_impl`, yet the same error appears (just without debug info). And this makes totally sense, loading dwarf data should not directly interfere with socket IO at all.

**Comment by straight-shoota** [comment]
As a quick fix I suppose we can revert https://github.com/crystal-lang/crystal/commit/6edc9e3920ab6e676bd1a27da084c8e1013cdfc1 for all platforms except FreeBSD. It's not very clean, but seemed to work before.

**Comment by beta-ziliani** [comment]
🙇 @Blacksmoke16 for the detailed description and research of the issue 🙇

**Comment by straight-shoota** [comment]
Through careful examination of the [massive LLVM IR diff](https://gist.github.com/straight-shoota/fdb6d7c03d4668d02c19d7c2abc12b48) with a highly optimized "good luck" approach that prioritizes on code variations in locations directly influenced by the commit in question ([lines 854f.](https://gist.github.com/straight-shoota/fdb6d7c03d4668d02c19d7c2abc12b48#file-regession-io-no-debug-ll-diff-L854-L855) immediately stood out), we were able to identify a possible remedy in an opportunistically inspired experimental approach.
The conjured patch is quite minimal:
```diff
--- i/src/exception/call_stack/dwarf.cr
+++ w/src/exception/call_stack/dwarf.cr
@@ -11,7 +11,7 @@ struct Exception::CallStack
   @@dwarf_function_names : Array(Tuple(LibC::SizeT, LibC::SizeT, String))?

   # :nodoc:
-  def self.load_debug_info
+  def self.load_debug_info : Nil
     return if ENV["CRYSTAL_LOAD_DEBUG_INFO"]? == "0"

     unless @@dwarf_loaded
```

Yet the reasons for the bug are still as unclear as how this patch fixes it.

Apparently the return type of `load_debug_info` has some kind of effect on `buffered_write`. Before https://github.com/crystal-lang/crystal/commit/6edc9e3920ab6e676bd1a27da084c8e1013cdfc1 the return type was `Int32 | Nil`, afterwards it's `Array(Tuple(UInt64, UInt64, String)) | Int32 | Nil` and with above patch it's `Nil` (which is correct as it doesn't actually return anything, it was just implicit return expressions).

**Comment by asterite** [comment]
> Apparently the return type of load_debug_info has some kind of effect on buffered_write

Where is `buffered_write` defined? I can't find it.

**Comment by straight-shoota** [comment]
Sorry, it's `unbuffered_write`, of course. The stack trace is not entirely conclusive, but it must be the def in `Crystal::System::Socket`. Due to performance optimizations, it seems several methods are mashed together. So I don't think we have established where exactly the actual `IndexError` is raised. Neither `src/crystal/spin_lock.cr:25:5` nor `Crystal::System::Socket#unbuffered_write` point to any kind of index lookup.

**Comment by jwoertink** [comment]
Just deployed 1.7.2 to production, but this is still happening in our app. Interesting enough, if we add the `content_length` on the long calls, it'll at least return the full data, it just throws the Index out of bounds along with it...

**Comment by dmgk** [comment]
Another possible workaround is to avoid printing exception backtraces if debug info failed to load:

```diff
diff --git a/src/exception/call_stack/dwarf.cr b/src/exception/call_stack/dwarf.cr
index 96d99f032..549f744a3 100644
--- a/src/exception/call_stack/dwarf.cr
+++ b/src/exception/call_stack/dwarf.cr
@@ -21,7 +21,7 @@ struct Exception::CallStack
       rescue ex
         @@dwarf_line_numbers = nil
         @@dwarf_function_names = nil
-        Crystal::System.print_exception "Unable to load dwarf information", ex
+        Crystal::System.print_error "Unable to load dwarf information"
       end
     end
   end
diff --git a/src/exception/call_stack/stackwalk.cr b/src/exception/call_stack/stackwalk.cr
index 73b1060aa..89639687a 100644
--- a/src/exception/call_stack/stackwalk.cr
+++ b/src/exception/call_stack/stackwalk.cr
@@ -15,7 +15,7 @@ struct Exception::CallStack
       begin
         load_debug_info_impl
       rescue ex
-        Crystal::System.print_exception "Unable to load debug information", ex
+        Crystal::System.print_error "Unable to load debug information"
       end
     end
   end

```

There's something going on during semantic analysis which I wasn't quite able to get to the bottom of in https://github.com/crystal-lang/crystal/issues/12329, but it seems to affect code generation.

**Comment by jwoertink** [comment]
@straight-shoota actually, it doesn't... I'm kind of surprised. I'm able to recreate the issue locally, at least. I pulled Crystal from master, reverted that single file, `make clean crystal`, then built the app in release mode and got the error.

```
▸  IndexError 

     Index out of bounds

    Backtrace 

     /home/jeremy/Development/crystal/lang/src/crystal/spin_lock.cr:25:5 in 'unbuffered_write'
     /home/jeremy/Development/crystal/lang/src/http/server/response.cr:263:11 in 'unbuffered_write'
     /home/jeremy/Development/crystal/lang/src/string.cr:5251:5 in 'print'
     lib/lucky/src/lucky/route_handler.cr:10:7 in 'call'
     /home/jeremy/Development/crystal/lang/src/http/server/request_processor.cr:51:11 in 'handle_client'
     /home/jeremy/Development/crystal/lang/src/fiber.cr:146:11 in 'run'
     ???
```

I'm down to try anything if anyone has some ideas.

**Comment by straight-shoota** [comment]
I don't think you need to rebuild the compiler. Eventually, there seems to be some kind of compiler bug involved, but there's no indication that any change to the compiler triggered it. Instead it looks like something changed in stdlib that through some side effect makes this error appear. Probably in a similar way as https://github.com/crystal-lang/crystal/pull/12992 what we have identified as a workaround for the isolated reproduction example (but apparently not for your full production code).

What you could do is doing a bisect on the changes in this repo between 1.6.2 and 1.7.0 to figure out which commit is the first one that breaks your production code.

**Comment by jwoertink** [comment]
Well, I spent several hours trying to figure this out, but I can't get an old enough version of Crystal to run locally in order to bisect it. The last version it worked on isn't 1.6.2, it's 1.1.1. So some commit after 1.1.1 is where it broke. The earliest version I was able to run locally is 1.3.0, and that still has the error. 

I also tried to get this to run in a Docker container, but I kept running in to all kinds of random issues like `Error: can't find file 'prelude'` and `undefined reference to `EVP_CIPHER_block_size'` and a few LLVM related issues. If someone is able to create a Dockerfile I can just drop in and run, I can test it, but I think this is beyond my skillset at this point.

**Comment by jwoertink** [comment]
Ok, scratch that... I may have found it....

Bisect says this is the culprit https://github.com/crystal-lang/crystal/pull/11003

With some help from @Blacksmoke16 I was able to run `sudo docker run --rm -it -v $PWD:/app -w /app crystallang/crystal:1.1.1-build bash` locally and bisect 

```
lang on  master [⇣] 
❯ git bisect start

lang on  master (BISECTING) [⇣] 
❯ git bisect good 6d9a1d5

lang on  master (BISECTING) [⇣] 
❯ git bisect bad 9f90efe
Bisecting: 97 revisions left to test after this (roughly 7 steps)
[ccd271cc5555535e555271f3179736eab72155e3] Primitive annotations for interpreter (#11147)

lang on  HEAD (ccd271c) (BISECTING) 
❯ git bisect bad
Bisecting: 47 revisions left to test after this (roughly 6 steps)
[32659bc7c6be76111afc3dc00ef4ac5fcee5806d] Merge pull request #11040 from straight-shoota/feature/markd

lang on  HEAD (32659bc) (BISECTING) 
❯ git bisect bad
Bisecting: 24 revisions left to test after this (roughly 5 steps)
[58e582a90a57cba0b43e6f951db8152be1e5148f] crystal tool format (#11006)

lang on  HEAD (58e582a) (BISECTING) 
❯ git bisect good
Bisecting: 12 revisions left to test after this (roughly 4 steps)
[4ad72d86e203b9f7667e65b666cb22c1b70c0310] Support basic auth from URI in websockets (#10854)

lang on  HEAD (4ad72d8) (BISECTING) 
❯ git bisect bad
Bisecting: 5 revisions left to test after this (roughly 3 steps)
[32f034fb9519fe0a7efb114829a5e4b7807699f8] Add precision parameter to `Int#to_s` (#10926)

lang on  HEAD (32f034f) (BISECTING) 
❯ git bisect good
Bisecting: 2 revisions left to test after this (roughly 2 steps)
[3cc36c9c7f53b20f99d76331e246d52594d9c5f5] Fix no overflow check when primitive int converts to same type (#11003)

lang on  HEAD (3cc36c9) (BISECTING) 
❯ git bisect bad
Bisecting: 0 revisions left to test after this (roughly 1 step)
[dd818e92d2ba95e989a16a3c31aec12403a5c625] Update readme to point to IRC channel on libera.chat (#11024)

lang on  HEAD (dd818e9) (BISECTING) 
❯ git bisect good
3cc36c9c7f53b20f99d76331e246d52594d9c5f5 is the first bad commit
commit 3cc36c9c7f53b20f99d76331e246d52594d9c5f5
Author: Quinton Miller <nicetas.c@gmail.com>
Date:   Tue Jul 27 17:04:13 2021 +0800

    Fix no overflow check when primitive int converts to same type (#11003)

 spec/compiler/codegen/primitives_spec.cr   | 4 ++++
 src/compiler/crystal/codegen/primitives.cr | 2 +-
 2 files changed, 5 insertions(+), 1 deletion(-)
```

However, after that was done, I wanted to test it, so I checked out the latest master commit and ran

`git revert 3cc36c9c7f53b20f99d76331e246d52594d9c5f5`, then `make clean && make`. After building the release version of my app, it still fails :confused:  Maybe it's possible that this bug is the culmination of several separate commits?

**Comment by straight-shoota** [comment]
>  Maybe it's possible that this bug is the culmination of several separate commits?

That may very well be the case.
A way to verify that and find other commits that are involved is to run bisect again. Now starting with 3cc36c9c7f53b20f99d76331e246d52594d9c5f5 as the last good commit and applying the revert patch on every commit you're testing.
I'd recommend using a script for that.
Something like this:
```sh
# bisect-helper.sh
git revert --no-commit 3cc36c9c7f53b20f99d76331e246d52594d9c5f5
your_test_command
exit_code=$?
git checkout -- .
exit exit_code
```
You can run it as `git bisect run ./bisect-helper.sh`.

**Comment by jwoertink** [comment]
Thanks. Now that I have a sandbox, it should be easier to dive in. I'll try that out when I get some free time.

**Comment by ggiraldez** [comment]
I stumbled upon this issue in an internal project, so I started investigating a bit.

## Short version

The tl;dr version is I found a lib C definition which is wrong. In this case, the signature for `send` in [socket.cr](https://github.com/crystal-lang/crystal/blob/master/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr#L59) types the third parameter as `Int` (signed 32-bit), but it's a `SizeT` (unsigned 64-bit) (see [send(2)](https://www.man7.org/linux/man-pages/man2/send.2.html)). I tried patching it locally for the binary distribution of Crystal 1.7.3 as well as in a local source build from the `master` branch and fixing that type seems to work. I suspect there may be many other instances of such mismatches.

So this patch:

```diff
diff --git a/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr b/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr
index 6cee17373..4503d20dd 100644
--- a/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr
+++ b/src/lib_c/x86_64-linux-gnu/c/sys/socket.cr
@@ -56,7 +56,7 @@ lib LibC
   fun listen(fd : Int, n : Int) : Int
   fun recv(fd : Int, buf : Void*, n : Int, flags : Int) : SSizeT
   fun recvfrom(fd : Int, buf : Void*, n : Int, flags : Int, addr : Sockaddr*, addr_len : SocklenT*) : SSizeT
-  fun send(fd : Int, buf : Void*, n : Int, flags : Int) : SSizeT
+  fun send(fd : Int, buf : Void*, n : SizeT, flags : Int) : SSizeT
   fun sendto(fd : Int, buf : Void*, n : Int, flags : Int, addr : Sockaddr*, addr_len : SocklenT) : SSizeT
   fun setsockopt(fd : Int, level : Int, optname : Int, optval : Void*, optlen : SocklenT) : Int
   fun shutdown(fd : Int, how : Int) : Int
```

## Long version

I compiled the program in release mode to trigger the issue and run it through a debugger. Since it's not a crash, it took me a bit to set some useful breakpoints. The first weird thing is that when run under a debugger the reported error is different:

```
(gdb) run
Starting program: /home/ggiraldez/Scratchpad/crystal-12982/test 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Unhandled exception in spawn: Error while writing data to the client (HTTP::Server::ClientError)
  from /home/ggiraldez/Projects/crystal/src/http/server/response.cr:249:9 in 'unbuffered_write'
  from test.cr:11:5 in '->'
  from /home/ggiraldez/Projects/crystal/src/fiber.cr:146:11 in 'run'
  from ???
Caused by: Error writing to socket: Bad address (IO::Error)
  from /home/ggiraldez/Projects/crystal/src/gc/boehm.cr:132:5 in 'unbuffered_write'
  from /home/ggiraldez/Projects/crystal/src/http/server/response.cr:263:11 in 'unbuffered_write'
  from test.cr:11:5 in '->'
  from /home/ggiraldez/Projects/crystal/src/fiber.cr:146:11 in 'run'
  from ???
[Inferior 1 (process 9291) exited normally]
(gdb)
```

We get a "bad address" error, ie. a segfault. But the fault probably happens in kernel mode, inside the `send` call. So `send` returns -1 and sets `errno` to 14 (EFAULT).  I'm not sure why the reported error is different if the program is run under a debugger. Anyway, I set a breakpoint for `send` and the I got this two hits:

```
(gdb) b send
Breakpoint 1 at 0x555555593b93: send. (6 locations)
(gdb) run
Starting program: /home/ggiraldez/Scratchpad/crystal-12982/test 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, __libc_send (fd=12, buf=0x7ffff6887000, len=53, flags=0)
    at ../sysdeps/unix/sysv/linux/send.c:24
24	../sysdeps/unix/sysv/linux/send.c: No such file or directory.
(gdb) p (char*)buf
$1 = 0x7ffff6887000 "HTTP/1.1 200 OK\r\nTransfer-Encoding: chunked\r\n\r\naff3\r\n"
(gdb) c
Continuing.

Breakpoint 1, __libc_send (fd=12, buf=0x7ffff78ac00c, len=4295012339, flags=0)
    at ../sysdeps/unix/sysv/linux/send.c:24
24	in ../sysdeps/unix/sysv/linux/send.c
(gdb) p (char*)buf
$2 = 0x7ffff78ac00c 'a' <repeats 200 times>...
(gdb) p/x len
$3 = 0x10000aff3
(gdb) 
```

The `len` parameter to `__libc_send` is wrong in the second call. The parameter is passed in a register, and that register's upper 32 bits contain some extra non-zero bits (presumably from a previous computation). Since `size_t` is 64 bits on 64-bit Linux, but `Int` is a 32-bit integer, the compiler incorrectly assumes the callee (lib C's `send` in this case) will ignore the most significant 32 bits.

Changing the type in the `lib` definition ensures the compiler properly extends (by zero padding) the `Slice#size` property into a `SizeT` before passing it to `send`. I later verified the generated assembly code for the calling code.

**Comment by bararchy** [comment]
@ggiraldez great find :rocket: I wonder where more we get this kind of issues, and it might explain all kind of other instabilities

**Comment by beta-ziliani** [comment]
We'll be looking into the bindings 🕵️ 
It'd be great to have an automatic check for such things...

**Comment by jwoertink** [comment]
wow.... :exploding_head: Yeah, I never would have found that lol. You are my savior @ggiraldez :raised_hands:
